src/test/java/org/sagebionetworks/bridge/spring/util/EtagComponentTest.java

```diff
@@ -0,0 +1,256 @@
+    public void nullArgThrowsException() throws Throwable {
+        when(mockRequest.getHeader(IF_NONE_MATCH)).thenReturn(ETAG);
+        
+        doReturn(ImmutableMap.of()).when(mockContext).getArgValues();
+        
+        component.checkEtag(mockJoinPoint);
+    }
+    
+    @Test
+    public void cacheMiss_timestampMissing() throws Throwable {
+        when(mockRequest.getHeader(IF_NONE_MATCH)).thenReturn(ETAG);
+        
+        CacheKey userKey = CacheKey.etag(Account.class, TEST_USER_ID);
+        when(mockCacheProvider.getObject(userKey, DateTime.class)).thenReturn(null);
+        
+        Object retValue = component.checkEtag(mockJoinPoint);
+        assertEquals(retValue, ACCOUNT_ID);
+    }
+    
+    // retrieving appId and userId are tested in the default set up for these tests.
+    
+    @Test
+    public void orgIdRetrievedFromSession() throws Throwable {
+        when(mockRequest.getHeader(IF_NONE_MATCH)).thenReturn(ETAG);
+        
+        EtagCacheKey orgKeyAnn = new EtagCacheKey() {
+            @Override
+            public Class<? extends Annotation> annotationType() {
+                return EtagCacheKey.class;
+            }
+            @Override
+            public Class<?> model() {
+                return Organization.class;
+            }
+            @Override
+            public String[] keys() {
+                return new String[] {"orgId"};
+            }
+        };
+        doReturn(ImmutableList.of(orgKeyAnn)).when(mockContext).getCacheKeys();
+        
+        CacheKey orgKey = CacheKey.etag(Organization.class, TEST_ORG_ID);
+        when(mockCacheProvider.getObject(orgKey, DateTime.class)).thenReturn(MODIFIED_ON);
+        
+        String stringToHash = MODIFIED_ON.toString();
+        when(mockMd5DigestUtils.digest(stringToHash.getBytes())).thenReturn("ETAG".getBytes());
+        
+        Object retValue = component.checkEtag(mockJoinPoint);
+        
+        assertNull(retValue);
+        verify(mockResponse).addHeader(HttpHeaders.ETAG, ETAG);
+        verify(mockResponse).setStatus(304);        
+    }
+    
+    // This will probably never trigger in reality, since the values are usually path parameters
+    // in the controller. But it's logically possible.
+    @Test(expectedExceptions = IllegalArgumentException.class,
+        expectedExceptionsMessageRegExp = "EtagSupport: no value for key: studyId")
+    public void shouldHaveProvidedArgumentValueButDidNot() throws Throwable {
+        when(mockRequest.getHeader(IF_NONE_MATCH)).thenReturn(ETAG);
+        
+        Map<String, Object> map = new HashMap<>();
+        map.put("studyId", null);
+        doReturn(map).when(mockContext).getArgValues();
+        
+        Object retValue = component.checkEtag(mockJoinPoint);
+        
+        assertNull(retValue);
+        verify(mockResponse).addHeader(HttpHeaders.ETAG, ETAG);
+        verify(mockResponse).setStatus(304);
```

Comment:
Minor: These asserts and verifies will never be called, if component.getEtag() throws.